### PR TITLE
GATE-2448: Adds support for proxy endpoints

### PR DESCRIPTION
**Pull request by Michael9127**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 FEATURES:
 
 * **New Resource:** `cloudflare_waiting_room_event` ([#1509](https://github.com/cloudflare/terraform-provider-cloudflare/issues/1509))
+* **New Resource:** `cloudflare_teams_proxy_endpoint`
```
Comment (jacobbednarz) [inline CHANGELOG.md:6]: this file is automatically generated, check out https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/docs/changelog-process.md to add a changelog entry.

File: cloudflare/resource_cloudflare_teams_proxy_endpoints.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	if err := d.Set("name", endpoint.Name); err != nil {
+		return fmt.Errorf("error parsing Proxy Endpoint name")
+	}
+	if err := d.Set("ips", endpoint.IPs); err != nil {
+		return fmt.Errorf("error parsing Proxy Endpoint IPs")
+	}
+	if err := d.Set("subdomain", endpoint.Subdomain); err != nil {
+		return fmt.Errorf("error parsing Proxy Endpoint subdomain")
+	}
+	return nil
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_proxy_endpoints.go:49]:
```suggestion
	if err := d.Set("name", endpoint.Name); err != nil {
		return fmt.Errorf("error parsing Proxy Endpoint name")
	}
	
	if err := d.Set("ips", endpoint.IPs); err != nil {
		return fmt.Errorf("error parsing Proxy Endpoint IPs")
	}
	
	if err := d.Set("subdomain", endpoint.Subdomain); err != nil {
		return fmt.Errorf("error parsing Proxy Endpoint subdomain")
	}
	
	return nil
```

File: cloudflare/resource_cloudflare_teams_proxy_endpoints.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	log.Printf("[DEBUG] Updating Cloudflare Teams Proxy Endpoint from struct: %+v", updatedProxyEndpoint)
+
+	teamsProxyEndpoint, err := client.UpdateTeamsProxyEndpoint(context.Background(), accountID, updatedProxyEndpoint)
+	if err != nil {
+		return fmt.Errorf("error updating Teams Proxy Endpoint for account %q: %s", accountID, err)
+	}
+	if teamsProxyEndpoint.ID == "" {
+		return fmt.Errorf("failed to find Teams Proxy Endpoint ID in update response; resource was empty")
+	}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_proxy_endpoints.go:87]:
```suggestion

	if err != nil {
		return fmt.Errorf("error updating Teams Proxy Endpoint for account %q: %s", accountID, err)
	}
	
	if teamsProxyEndpoint.ID == "" {
		return fmt.Errorf("failed to find Teams Proxy Endpoint ID in update response; resource was empty")
	}
```

File: cloudflare/resource_cloudflare_teams_proxy_endpoints.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+}
+func resourceCloudflareTeamsProxyEndpointUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	accountID := d.Get("account_id").(string)
+	updatedProxyEndpoint := cloudflare.TeamsProxyEndpoint{
+		ID:   d.Id(),
+		Name: d.Get("name").(string),
+		IPs:  inflateProxyEndpointIPs(d.Get("ips").([]interface{})),
+	}
+	log.Printf("[DEBUG] Updating Cloudflare Teams Proxy Endpoint from struct: %+v", updatedProxyEndpoint)
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_proxy_endpoints.go:79]:
```suggestion

	log.Printf("[DEBUG] Updating Cloudflare Teams Proxy Endpoint from struct: %+v", updatedProxyEndpoint)
```

File: cloudflare/resource_cloudflare_teams_proxy_endpoints_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudflareTeamsProxyEndpointDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCloudflareTeamsProxyEndpointConfigBasic(rnd, accountID),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(name, "account_id", accountID),
+					resource.TestCheckResourceAttr(name, "name", rnd),
+					resource.TestCheckResourceAttr(name, "ips.0", "1.2.3.4/32"),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_proxy_endpoints_test.go:41]: let's use IP ranges from RFC5737 for our tests/documentation.

File: cloudflare/schema_cloudflare_teams_proxy_endpoints.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+		"name": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"subdomain": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"ips": {
+			Type:     schema.TypeList,
```
Comment (jacobbednarz) [inline cloudflare/schema_cloudflare_teams_proxy_endpoints.go:21]: do you care about order? if not, you probably want `TypeSet`.

**Review by jacobbednarz — CHANGES_REQUESTED** [review]
solid start! to get this over the line we need:

- a CHANGELOG entry; and
- website documentation. check out #1113 for a good example.

File: cloudflare/schema_cloudflare_teams_proxy_endpoints.go (hunk 6 above)
Comment (Michael9127) [reply cloudflare/schema_cloudflare_teams_proxy_endpoints.go]: nice, done

File: cloudflare/resource_cloudflare_teams_proxy_endpoints_test.go (hunk 5 above)
Comment (Michael9127) [reply cloudflare/resource_cloudflare_teams_proxy_endpoints_test.go:41]: ah right, I forget every time. Fixed!

File: CHANGELOG.md (hunk 1 above)
Comment (Michael9127) [reply CHANGELOG.md]: fixed!

File: cloudflare/resource_cloudflare_teams_proxy_endpoints.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	return resourceCloudflareTeamsProxyEndpointRead(d, meta)
+
+}
+func resourceCloudflareTeamsProxyEndpointUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	accountID := d.Get("account_id").(string)
+	updatedProxyEndpoint := cloudflare.TeamsProxyEndpoint{
+		ID:   d.Id(),
+		Name: d.Get("name").(string),
+		IPs:  inflateProxyEndpointIPs(d.Get("ips").([]interface{})),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_proxy_endpoints.go:80]:
```suggestion
		IPs:  expandInterfaceToStringList(d.Get("ips").([]interface{})),
```

File: cloudflare/resource_cloudflare_teams_proxy_endpoints.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+}
+
+func inflateProxyEndpointIPs(ips []interface{}) []string {
+	i := make([]string, len(ips))
+	for x, ip := range ips {
+		i[x] = ip.(string)
+	}
+	return i
+}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_proxy_endpoints.go:138]: duplicates `expandInterfaceToStringList`

```suggestion
```

**Comment by jacobbednarz** [comment]
acceptance tests are passing 

```
TF_ACC=1 go test $(go list ./...) -v -run "^TestAccCloudflareTeamsProxyEndpoint_" -count 1 -parallel 1 -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestAccCloudflareTeamsProxyEndpoint_Basic
--- PASS: TestAccCloudflareTeamsProxyEndpoint_Basic (10.54s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	11.092s
?   	github.com/cloudflare/terraform-provider-cloudflare/tools/cmd/changelog-check	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/tools/cmd/maintainer-only-file-check	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/tools/cmd/tf-log-check	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
```

**Comment by jacobbednarz** [comment]
thank you @Michael9127 👏

**Pull request merged**